File: files/300-399/356/CH-356.py
import pandas as pd
path = "300-399/356/CH-356 Table Transformation.xlsx"
input = pd.read_excel(path, usecols="B", skiprows=2, nrows=15)
test = pd.read_excel(path, usecols="D:F", skiprows=2, nrows=5)

input['Type'] = input['Name'].apply(lambda x: type(x).__name__)
input = input.sort_values('Type').reset_index(drop=True)
input['GroupIndex'] = input.groupby('Type').cumcount() + 1
input['Name'] = input.apply(lambda row: pd.Timestamp(row['Name']) if type(row['Name']).__name__ == 'datetime' else row['Name'], axis=1)
pivoted = input.pivot(index='GroupIndex', columns='Type', values='Name')
pivoted['int'] = pivoted['int'].astype('int64')
column_map = {'datetime': 'Date', 'str': 'Product', 'int': 'Sale'}
pivoted = pivoted.rename(columns=column_map)
desired_order = ['Date', 'Product', 'Sale']
pivoted = pivoted[[col for col in desired_order if col in pivoted.columns]].reset_index(drop=True)
pivoted.columns.name = None

print(all(pivoted == test))
# True

File: files/300-399/356/CH-356.R
library(tidyverse)
library(readxl)
library(janitor)

path <- "300-399/356/CH-356 Table Transformation.xlsx"
input <- read_excel(path, range = "B3:B18")
test <- read_excel(path, range = "D3:F8") %>%
  mutate(Date = as.Date(Date, origin = "1899-12-30"))

Date <- input %>%
  filter(str_length(Name) > 3) %>%
  mutate(Date = excel_numeric_to_date(as.numeric(Name))) %>%
  select(Date)
Product <- input %>%
  filter(str_detect(Name, "[A-Za-z]")) %>%
  rename(Product = Name)
Sale <- input %>%
  filter(str_detect(Name, "^[0-9]{1,2}$")) %>%
  mutate(Sale = as.numeric(Name)) %>%
  select(Sale)

final <- bind_cols(Date, Product, Sale)
all_equal(final, test)
# [1] TRUE
